File: SWE_eclips/src/algorithm/BOJ/dp/LIS.java
package algorithm.BOJ.dp;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class LIS {
	 public static void main(String[] args) throws IOException {

	BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    StringTokenizer st;

    int N = Integer.parseInt(br.readLine());
    int[] arr = new int[N + 1];
    // 자신을 끝으로 하는 LIS 최장길이
    int[] LIS = new int[N + 1];

    st = new StringTokenizer(br.readLine());

    for (int i = 1; i <= N; i++) {
        arr[i] = Integer.parseInt(st.nextToken());
    }

    int res = 0;
    for (int i = 1; i <= N; i++) {
        // 자신만으로 LIS 구성했을 때의 길이는 1
        LIS[i] = 1;

        // 자신(i)의 앞에 있는 원소들과 비교
        for (int j = 0; j < i; j++) {
            // 앞쪽 원소보다 자신이 큰 경우
            if (arr[j] < arr[i] && LIS[i] < LIS[j] + 1)
                LIS[i] = LIS[j] + 1;
        }
        // 현재 원소의 LIS 값과 전체 최대값하고 비교하여 최댓값 갱신
        res = Math.max(res, LIS[i]);
    }

    System.out.println(res);

}


}

File: Theory/dp/lis.py
n = int(input())
arr = list(map(int,input().split()))
arr.reverse()

# 1 로 초기화 하는 이유는 적어도 자기 자신은 존재하기 때문.
dp = [1] * n

for i in range(1,n):
# dp[i] = i 를 마지막 원소로 가지는 부분 수열의 최대 길이
  for j in range(0, i):
  # 0~j까지의 dp를 살피며 arr[i]가 더 크면 dp[j] + 1
  # + 1은 arr[i]를 포함해야 하기 때문임.
    if arr[j] < arr[i]:
      dp[i] = max(dp[i],dp[j] +1)
# max(dp) = lis.
print(n - max(dp)) 